### ISSUE TITLE
CPP fragments in Skylark rule does not provide all fields

### ISSUE THREAD

**Issue by jbasila** [bazelbuild/bazel#3558]
### Description of the problem / feature request / question:
When using Skylark rules with 'cpp' fragments, not all fields that are described in the documentation are accessible.

Here is a simple example:
**rules.bzl**
```
def _my_rule_impl(ctx):                                    
    print(ctx.fragments.cpp.ar_executable)                 
    print(ctx.fragments.cpp.built_in_include_directories)  
    print(ctx.fragments.cpp.c_options)                     
    print(ctx.fragments.cpp.compiler)                      
    print(ctx.fragments.cpp.compiler_executable)           
    # print(ctx.fragments.cpp.compiler_options)            
    print(ctx.fragments.cpp.cpu)                           
    # print(ctx.fragments.cpp.cxx_options)                 
    # print(ctx.fragments.cpp.dynamic_link_options)        
    # print(ctx.fragments.cpp.fully_static_link_options)   
    print(ctx.fragments.cpp.libc)                          
    print(ctx.fragments.cpp.link_options)                  
    # print(ctx.fragments.cpp.mostly_static_link_options)  
    print(ctx.fragments.cpp.nm_executable)                 
    print(ctx.fragments.cpp.objcopy_executable)            
    print(ctx.fragments.cpp.objdump_executable)            
    print(ctx.fragments.cpp.preprocessor_executable)       
    print(ctx.fragments.cpp.strip_executable)              
    # print(ctx.fragments.cpp.sysroot)                     
    print(ctx.fragments.cpp.target_gnu_system_name)        
    # print(ctx.fragments.cpp.unfiltered_compiler_options) 

my_rule = rule(                                            
    implementation = _my_rule_impl,                        
    fragments = [ 'cpp' ],                                 
)
```

**BUILD**
```
load(':rule.bzl', 'my_rule')

my_rule(
    name = 'testing',
)
```

All the commented out lines in **rules.bzl** do not function.

### Environment info
* Operating System: Linux

* Bazel version (output of `bazel info release`):
I tried with:

1. release 0.5.3
2. development version (24624776cc23dfacac20eb554ccefe268f3b1a0a)

**Comment by iirina** [comment]
> All the commented out lines in rules.bzl do not function.

What does it mean "do not function"? They don't output anything or raise an error or something else?

**Comment by jbasila** [comment]
You get an error that the field does not exist in the cop fragment.

It turns out that compiler_options can be used but the syntax of:
Ctx.fragments.cpp.compiler_options([])
Not sure where this is documented that you need the ([]) to make this work - got this info from stack overflow

**Comment by iirina** [comment]
You are right, this should be documented and also have a default value (empty list) for the parameters. Thanks for reporting!

**Comment by jbasila** [comment]
Do please note that `compiler_options` is a specific one that I did manage to use as well as `cxx_options`, but the rest of the cpp fragment attributes that I commented out I still am unable to use and I get a generic error of (as an example I'm taking the `dynamic_link_options:
`object of type 'cpp' has no field 'dynamic_link_options'.`

If I use `ctx.fragments.cpp.dynamic_link_options([])` I get:
`too many arguments, in method dynamic_link_options(list) of 'cpp'.`

**Comment by iirina** [comment]
I understand. Until we document this, can you please take a look at our [source code](https://github.com/bazelbuild/bazel/blob/0abf5fa2d64c76def5a8fa0f960b73ce0566af4d/src/main/java/com/google/devtools/build/lib/rules/cpp/CppConfiguration.java) and look for each option that you cannot use and see what arguments does it take? It should unblock you for now, but we do need to document this ASAP.

**Comment by dslomov** [comment]
Assigning to @mhlopko, C++ expert

**Comment by jmillikin** [comment]
Not only do the docs not mention the required params, but the error message from Bazel is very confusing:

```
Traceback (most recent call last):
	File "/Users/john/src/rules_haskell/cc_test/BUILD", line 98
		_impl_debug_2(name = 'debug')
	File "/Users/john/src/rules_haskell/cc_test/debug.bzl", line 4, in _debug
		print(ctx.fragments.cpp.compiler_optio...())
	File "/Users/john/src/rules_haskell/cc_test/debug.bzl", line 4, in print
		ctx.fragments.cpp.compiler_options()
too many arguments, in method compiler_options() of 'cpp'
```

Bazel's "too many arguments" implies that _I'm_ providing too many arguments, when actually the error is _not enough_ arguments.

**Comment by mhlopko** [comment]
Still in the US meeting with people so pardon low service.
@jbasila you mention fields mentioned in documentation, what documentation do you mean? Is there any?

@jmillikin I'll take a look on the error message.

All in all, this API is going away very soon, O(weeks). That's the main reason it's not documented, we just don't want people to be using it :) It will be replaced by something else. but I still haven't started collecting the requirements so I cannot tell you what that something else will look like. If you describe your use case, I can consider it in the design.

**Comment by jbasila** [comment]
This is the documentation I used.
https://docs.bazel.build/versions/master/skylark/lib/cpp.html

**Comment by mhlopko** [comment]
Oh I see, so the functions that didn't work accept arguments (undocumented, as you already discovered). This is where all those functions are defined: https://github.com/bazelbuild/bazel/blob/275ae45b1228bdd0f912c4fbd634b29ba4180383/src/main/java/com/google/devtools/build/lib/rules/cpp/CppConfiguration.java if you need something quick and dirty. But again, the API is going away. The first step will very likely be `ctx.cpp.built_in_include_directories -> ctx.toolchains[//tools/cpp:toolchain_type].built_in_include_directories` refactoring, then the actual API will change.

**Comment by jmillikin** [comment]
> If you describe your use case, I can consider it in the design.

I'm making an attempt at `rules_haskell`. The Haskell compiler (GHC) works by generating assembly, then internally invoking `gcc` and `ld` to get object code and binaries. Telling it to stop after the assembly stage lets me use Bazel's compiler and linker options , which is better for hermeticity and consistency.

Currently I'm using the CPP fragment to discover the C++ toolchain, and `ctx.actions.execute` to compile. I'm hoping that can eventually be replaced by C++ sandwich providers, so that `cc_library -> hs_library -> cc_binary` becomes possible.

**Comment by mhlopko** [comment]
That will directly be supported by C++ sandwich \o/. @jbasila what was your usecase?

**Comment by jbasila** [comment]
The C++ sandwich is exactly what I need - I had to build a C++ library using Skylark but then I couldn't use it in other cc_library rules - I also use the cc_binary hack of libname.so with linkshared to create dynamic libraries the way I want them but then they cannot be used in the deps of other rules - again the sandwich should allow this.

**Comment by jbasila** [comment]
I need the CPP fragments for extracting compiler flags and such to be used in the Skylark rules such as the sysroot and compiler options

**Comment by mhlopko** [comment]
If we allowed you to create all the compilation and linking actions that cc_library does, and gave you an API to do so from Skylark, would that work for you too? In other words, instead of you taking the flags and creating actions yourselves, you would pass data to something like cc_common.compile/cc_common.link, and get some providers back, will that work?

**Comment by jmillikin** [comment]
You're thinking something like rules_go's [toolchain object](https://github.com/bazelbuild/rules_go/blob/master/go/toolchains.rst#the-toolchain-object)? That would work, but you'd need to give the caller more control over copts/linkopts than we can get with native CC rules today.

For Haskell rules:
* I need to inspect cc_options to see if PIC mode is active before building the Haskell compiler's argv.
* I need to drop some linker options, and add in a bunch more in a specific order.

**Comment by mhlopko** [comment]
That would be a next step, we will start with a higher level api that will encapsulate all action generation (so one call to `cc_common.compile` will register multiple actions depending on the state of almost all of the universe :).
PIC should be doable, linker options you're saying add in a specific order or also specific position? If latter, be prepared to do some crosstooling.

**Comment by jmillikin** [comment]
If your new API won't support fully specifying the options, can you keep the existing "c++ fragment" API around until it does? It would be very frustrating if existing rules written according to documented behavior stop working due to an internal refactoring.

For reference, here's the link step of rules_haskell. The order being constructed is: C main() stub, Haskell main, linker inputs (`-lfoo` etc), Bazel's calculated link options _without_ `-lstdc++`, and then a few more options that Haskell seems to need. The last part could maybe be moved up, but avoiding the link against `libstdc++` is pretty important.

I don't want to write a custom CROSSTOOL, it would be a horrible experience for users of these rules if they have to reconfigure their C/C++ compilations to use them.

```python
    bazel_link_options = _strip_libstdcxx(
          ctx.fragments.cpp.link_options
          + ctx.fragments.cpp.mostly_static_link_options(ctx.features, is_shared_lib))
    ctx.actions.run(
        outputs=[out_bin],
        inputs = depset([out_o]) + toolchain._ghc_private.rts_main.cc.libs + link_inputs,
        executable=ctx.fragments.cpp.compiler_executable,
        arguments = ([
            "-o",
            out_bin.path,
        ] + [l.path for l in toolchain._ghc_private.rts_main.cc.libs] + [
            out_o.path,
        ] + list(link_options)
          + bazel_link_options
          + [
            "-Wl,--gc-sections",
            "-Wl,-no-as-needed",
            "-Wl,-z,relro,-z,now",
        ]),
        mnemonic="HsBinary",
        progress_message="Linking %s to create [%s]" % (out_o, out_bin),
    )
```

**Comment by mhlopko** [comment]
The existing api is not correct, it doesn't work with feature configuration at all (so with the windows crosstool it returns empty lists everywhere), and its removal will not be because of some internal refactoring, but a cautious focused decision to remove it :) We'll share design docs, please make sure your use case is covered by the design.

Bazel will neither parse flags, nor will attach any semantics infomation on top of the commandlines generated from crosstool. If you need a particular flag present/not present in the command line, you need to have support in the crosstool. Of course I'm open to extending our builtin crosstools so haskel rules are supported out of the box, but if somebody created their own crosstool, they need to make sure it works with the rules.

**Comment by steeve** [comment]
hey @mhlopko, was there any work done on that front? I'm confronted with the very same issue for `rules_go` at https://github.com/bazelbuild/rules_go/blob/master/go/private/rules/stdlib.bzl#L59

Basically the stdlib is bootstrapped using Go itself and passing `CC` `CFLAGS` env vars and the like. However since the flags are incomplete, it fails to bootstrap when doing cross compilation (ie use apple_binary).
Basically it's missing `-isysroot`, for instance.

**Comment by steeve** [comment]
Actually, I'm thinking a good API would be to just get the tool and flagset for a certain action from the CROSSTOOL.

Something like:
```
ctx.crosstool.action("c-compile").{tool,flags,env}
```

**Comment by steeve** [comment]
For now I'm thinking we can parse the current CROSSTOOL via a homemade tool and expose this

**Comment by steeve** [comment]
Basically exposing `FeatureConfiguration.get{ToolForAction,CommandLine,EnvironmentVariables}` to Skylark would be perfect

**Comment by mhlopko** [comment]
Something like this? https://docs.google.com/document/d/1g91BWJITcYw_X-VxsDC0VgUn5E9g0kRBGoBSpoO41gA/edit

**Comment by mhlopko** [comment]
There's `get_tool_for_action`, isn't that what you need? Also feel free to comment directly on the doc.

**Comment by mhlopko** [comment]
The toolchain api is [done](https://github.com/bazelbuild/bazel/issues/4571), I'm going to close this issue now. Feel free to ask questions and file issues when you discover bugs.

Docs:
https://docs.bazel.build/versions/master/skylark/lib/cc_common.html
related docs to the cc toolchain info:
https://docs.bazel.build/versions/master/skylark/lib/CcToolchainInfo.html

The api on the CppConfiguration fragment is deprecated and will be soon removed.
